#! /bin/sh /usr/share/dpatch/dpatch-run
## Selected Text is not replaced by Input text when Japanese IM is on.dpatch by  <root@debian.example.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Patch by Alissa Sabre

@DPATCH@

Index: linden/indra/llui/lllineeditor.cpp
===================================================================
--- linden/indra/llui/lllineeditor.cpp	(revision 535)
+++ linden/indra/llui/lllineeditor.cpp	(revision 536)
@@ -2315,14 +2315,20 @@
 
 void LLLineEditor::resetPreedit()
 {
-	if (hasPreeditString())
+	if (hasSelection())
 	{
-		if (hasSelection())
+		if (hasPreeditString())
 		{
 			llwarns << "Preedit and selection!" << llendl;
 			deselect();
 		}
-
+		else
+		{
+			deleteSelection();
+		}
+	}
+	if (hasPreeditString())
+	{
 		const S32 preedit_pos = mPreeditPositions.front();
 		mText.erase(preedit_pos, mPreeditPositions.back() - preedit_pos);
 		mText.insert(preedit_pos, mPreeditOverwrittenWString);
Index: linden/indra/llui/lltexteditor.cpp
===================================================================
--- linden/indra/llui/lltexteditor.cpp	(revision 535)
+++ linden/indra/llui/lltexteditor.cpp	(revision 536)
@@ -4333,14 +4333,20 @@
 
 void LLTextEditor::resetPreedit()
 {
-	if (hasPreeditString())
+	if (hasSelection())
 	{
-		if (hasSelection())
+		if (hasPreeditString())
 		{
 			llwarns << "Preedit and selection!" << llendl;
 			deselect();
 		}
-
+		else
+		{
+			deleteSelection(FALSE);
+		}
+	}
+	if (hasPreeditString())
+	{
 		mCursorPos = mPreeditPositions.front();
 		removeStringNoUndo(mCursorPos, mPreeditPositions.back() - mCursorPos);
 		insertStringNoUndo(mCursorPos, mPreeditOverwrittenWString);
